dask_jobqueue/job.py

```diff
@@ -0,0 +1,480 @@
+            )
+        return out
+
+
+class JobQueueCluster(SpecCluster):
+    __doc__ = """ Deploy Dask on a Job queuing system
+
+    This is a superclass, and is rarely used directly.  It is more common to
+    use an object like SGECluster, SLURMCluster, PBSCluster, LSFCluster, or
+    others.
+
+    However, it can be used directly if you have a custom ``Job`` type.
+    This class relies heavily on being passed a ``Job`` type that is able to
+    launch one Job on a job queueing system.
+
+    Parameters
+    ----------
+    Job : Job
+        A class that can be awaited to ask for a single Job
+    {cluster_parameters}
+    """.format(
+        cluster_parameters=cluster_parameters
+    )
+
+    job_cls = None
+
+    def __init__(
+        self,
+        n_workers=0,
+        job_cls: Job = None,
+        # Cluster keywords
+        loop=None,
+        security=None,
+        silence_logs="error",
+        name=None,
+        asynchronous=False,
+        # Scheduler keywords
+        interface=None,
+        host=None,
+        protocol="tcp://",
+        dashboard_address=":8787",
+        config_name=None,
+        # Job keywords
+        **kwargs
+    ):
+        self.status = "created"
+        if job_cls is not None:
+            self.job_cls = job_cls
+
+        if self.job_cls is None:
+            raise ValueError(
+                "You must provide a Job type like PBSJob, SLURMJob, "
+                "or SGEJob with the job_cls= argument."
+            )
+
+        if config_name:
+            if interface is None:
+                interface = dask.config.get("jobqueue.%s.interface" % config_name)
+
+        scheduler = {
+            "cls": Scheduler,  # Use local scheduler for now
+            "options": {
+                "protocol": protocol,
+                "interface": interface,
+                "host": host,
+                "dashboard_address": dashboard_address,
+                "security": security,
+            },
+        }
+        if config_name:
+            kwargs["config_name"] = config_name
+        kwargs["interface"] = interface
+        kwargs["protocol"] = protocol
+        kwargs["security"] = security
+        self._kwargs = kwargs
+        worker = {"cls": self.job_cls, "options": kwargs}
+        if "processes" in kwargs and kwargs["processes"] > 1:
+            worker["group"] = ["-" + str(i) for i in range(kwargs["processes"])]
+
+        self.example_job  # trigger property to ensure that the job is valid
+
+        super().__init__(
+            scheduler=scheduler,
+            worker=worker,
+            loop=loop,
+            silence_logs=silence_logs,
+            asynchronous=asynchronous,
+            name=name,
+        )
+
+        if n_workers:
+            self.scale(n_workers)
+
+    @property
+    def example_job(self):
+        try:
+            address = self.scheduler.address
+        except AttributeError:
+            address = "tcp://scheduler:8786"
```

Comment:
I don't really understand the case where `self.scheduler.address` does not exist ...

Reply:
When we first construct the cluster, and have only called `__init__` we have not yet awaited the scheduler, so the address is not yet known.